src/map/map.c

```diff
@@ -3240,97 +3239,107 @@ int map_eraseipport(unsigned short map_index, uint32 ip, uint16 port) {
 	return 0;
 }
 
-/*==========================================
- * [Shinryo]: Init the mapcache
- *------------------------------------------*/
-char *map_init_mapcache(FILE *fp) {
-	struct map_cache_main_header header;
-	size_t size = 0;
-	char *buffer;
-
-	// No file open? Return..
-	nullpo_ret(fp);
+int map_readfromcache(struct map_data *m) {
+	unsigned int file_size;
+	char file_path[256];
+	FILE *fp = NULL;
+	int retval = 0;
+	int16 version;
+	
+	nullpo_ret(m);
+	
+	snprintf(file_path, sizeof(file_path), "%s%s%s.%s", "maps/", DBPATH, m->name, "mcache");
+	fp = fopen(file_path, "rb");
 
-	// Get file size
+	if (fp == NULL) {
+		ShowWarning("map_readfromcache: Could not find the file or read it for map '%s' path '%s'\n", m->name, file_path);
+		return 0;
+	}
+	
 	fseek(fp, 0, SEEK_END);
-	size = ftell(fp);
+	file_size = (unsigned int)ftell(fp);
 	fseek(fp, 0, SEEK_SET);
 
-	// Allocate enough space
-	CREATE(buffer, char, size);
-
-	// No memory? Return..
-	nullpo_ret(buffer);
-
-	// Read file into buffer..
-	if(fread(buffer, sizeof(char), size, fp) != size) {
-		ShowError("map_init_mapcache: Could not read entire mapcache file\n");
-		aFree(buffer);
-		return NULL;
+	if (fread(&version, sizeof(version), 1, fp) < 1) {
+		ShowError("map_readfromcache: Could not read file version for map '%s'\n", m->name);
+		fclose(fp);
+		return 0;
 	}
 
-	rewind(fp);
+	fseek(fp, 0, SEEK_SET);
 
-	// Get main header to verify if data is corrupted
-	if( fread(&header, sizeof(header), 1, fp) != 1 ) {
-		ShowError("map_init_mapcache: Error obtaining main header!\n");
-		aFree(buffer);
-		return NULL;
-	}
-	if( GetULong((unsigned char *)&(header.file_size)) != size ) {
-		ShowError("map_init_mapcache: Map cache is corrupted!\n");
-		aFree(buffer);
-		return NULL;
+	switch(version) {
+		case 1:
+			retval = map->readfromcache_v1(fp, m, file_size);
+			break;
+		default:
+			ShowError("map_readfromcache: unknow version passed for map '%s' version '%d'\n", m->name, version);
+			break;
 	}
 
-	return buffer;
+	fclose(fp);
+	return retval;
 }
 
-/*==========================================
- * Map cache reading
- * [Shinryo]: Optimized some behaviour to speed this up
- *==========================================*/
-int map_readfromcache(struct map_data *m, char *buffer) {
-	int i;
-	struct map_cache_main_header *header = (struct map_cache_main_header *)buffer;
-	struct map_cache_map_info *info = NULL;
-	char *p = buffer + sizeof(struct map_cache_main_header);
-
-	nullpo_ret(m);
-	nullpo_ret(buffer);
-
-	for(i = 0; i < header->map_count; i++) {
-		info = (struct map_cache_map_info *)p;
+int map_readfromcache_v1(FILE *fp, struct map_data *m, unsigned int file_size)
+{
+	struct map_cache_header mheader = {0};
+	uint8 md5buf[16] = {0};
+	int map_size;
+	nullpo_retr(false, fp);
+	nullpo_retr(false, m);
+
+	if (file_size < sizeof(mheader)) {
```

Comment:
Here should be check of <= sizeof(mheader) because at least one cell should be in map